File: 5. Week 5 - Calendar Heatmap/data/extract_data.py
import pandas as pd
import os
import numpy as np

def extract_data():
    input_path = '../../data/runs_only_redacted.csv'
    output_path = 'weekly_distance_heatmap.csv'
    
    if not os.path.exists(input_path):
        print(f"Error: Input file")
        return

    try:
        df = pd.read_csv(input_path)
        df.columns = [c.lower() for c in df.columns]
        
        # Parse dates
        if 'start_date' in df.columns:
            df['Date'] = pd.to_datetime(df['start_date'])
        else:
            return

        # Distance km
        df['Distance'] = df['distance'] / 1000
        
        # Calculate Pace?
        # Requirement: "tooltip showing total distance, avg pace, total runs"
        # Need moving_time
        if 'moving_time' in df.columns:
             df['Time'] = df['moving_time']
        elif 'average_speed' in df.columns:
             df['Time'] = df['distance'] / df['average_speed']
        else:
             df['Time'] = df['Distance'] * 5 * 60 # varying fallback? 
        
        # Extract Year and Week
        # Iso calendar is best for heatmaps usually
        df['Year'] = df['Date'].dt.isocalendar().year
        df['Week'] = df['Date'].dt.isocalendar().week
        
        # Aggregate
        weekly = df.groupby(['Year', 'Week']).agg(
            TotalDistance=('Distance', 'sum'),
            TotalTime=('Time', 'sum'),
            RunCount=('Distance', 'count')
        ).reset_index()
        
        # Avg Pace
        weekly['AveragePace'] = (weekly['TotalTime'] / 60) / weekly['TotalDistance']
        # Handle nan/inf
        weekly['AveragePace'] = weekly['AveragePace'].replace([np.inf, -np.inf], 0).fillna(0)
        
        # Round
        weekly['TotalDistance'] = weekly['TotalDistance'].round(2)
        weekly['AveragePace'] = weekly['AveragePace'].round(2)
        
        weekly.to_csv(output_path, index=False)
        print(f"Created {output_path}")

    except Exception as e:
        print(f"Error: {e}")

if __name__ == "__main__":
    extract_data()

File: 5. Week 5 - Calendar Heatmap/data/extract_data.R
# Week 5 - Data Extraction (R version)
# Goal: Aggregate data by Year and Week Number for Heatmap

library(tidyverse)
library(lubridate)

input_path <- "../../data/runs_only_redacted.csv"
output_path <- "weekly_distance_heatmap.csv"

if (file.exists(input_path)) {
  df <- read_csv(input_path, show_col_types = FALSE)
  
  heatmap_data <- df %>%
    mutate(Date = as.Date(Date)) %>%
    # ISOWek is often better for standards, but 'week' is fine.
    # Python script used 'isocalendar'.
    mutate(
      Year = isoyear(Date),
      Week = isoweek(Date)
    ) %>%
    group_by(Year, Week) %>%
    summarise(
      TotalDistance = sum(Distance, na.rm = TRUE) / 1000, # Convert to km
      RunCount = n(),
      .groups = "drop"
    ) %>%
    mutate(TotalDistance = round(TotalDistance, 2)) %>%
    arrange(Year, Week)
    
  write_csv(heatmap_data, output_path)
  print(paste("Saved data to", output_path))
  print(head(heatmap_data))

} else {
  stop("Input file not found")
}
